jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DebugHandler.java

```diff
@@ -30,16 +30,64 @@
 
 /**
  * Debug Handler.
+ * <p>
  * A lightweight debug handler that can be used in production code.
  * Details of the request and response are written to an output stream
  * and the current thread name is updated with information that will link
  * to the details in that output.
+ * </p>
+ * <p>
+ * Note that due to async processing, the logging of request processing may
+ * appear out of order.
+ * </p>
  */
 public class DebugHandler extends Handler.Wrapper implements Connection.Listener
 {
-    private final DateCache _date = new DateCache("HH:mm:ss", Locale.US);
+    private static final DateCache __date = new DateCache("yyyy-MM-dd HH:mm:ss", Locale.ENGLISH);
     private OutputStream _out;
     private PrintStream _print;
+    private boolean _showHeaders;
+    private final String _attr = String.format("__R%s@%x", this.getClass().getSimpleName(), System.identityHashCode(this));
+
+    private class HandlingCallback extends Callback.Nested
+    {
+        private final Request _request;
+        private final Response _response;
+        private final AtomicBoolean _handling = new AtomicBoolean(true); //when created, the request is being handled
+
+        private HandlingCallback(Callback callback, Request request, Response response)
+        {
+            super(callback);
+            _request = request;
+            _response = response;
+        }
+
+        private boolean handlingCompleted()
+        {
+            //test if the request is still being handled or not
+            return !_handling.compareAndSet(true, false);
+        }
+
+        private void callbackCompleted(Throwable throwable)
+        {
+            if (_handling.compareAndSet(true, false))
+                log("<< r=%s async=false %d %s%n%s", findRequestName(_request), _response.getStatus(), (throwable == null ? "" : throwable.toString()), _response.getHeaders());
+        }
```

Comment:
javadoc on these methods would be helpful

Reply:
I've changed the implementation to a) be more correct b) clearer and c) added more javadoc.